firedrake/mesh.py

```diff
@@ -2351,7 +2361,28 @@ def _pic_swarm_in_plex(plex, coords, fields=[]):
     # coordinates associated with them. The DMPlex cell id associated
     # with each PIC in the DMSwarm is accessed with the `DMSwarm_cellid`
     # field.
-    swarm.setPointCoordinates(coords, redundant=False, mode=PETSc.InsertMode.INSERT_VALUES)
+    # swarm.setPointCoordinates(coords, redundant=False, mode=PETSc.InsertMode.INSERT_VALUES)
+
+    # do this ourselves by just setting DMSwarmPICField_coor and DMSwarmPICField_cellid
+
+    # get fields - NOTE this isn't copied so make sure
+    # swarm.restoreField is called for each field too!
+    swarm_coords = swarm.getField("DMSwarmPIC_coor").reshape((num_vertices, gdim))
+    # Why are the next two different?
+    swarm_parent_cell_nums = swarm.getField("DMSwarm_cellid")
+    field_parent_cell_nums = swarm.getField("parentcellnum")
+    field_reference_coords = swarm.getField("refcoord").reshape((num_vertices, tdim))
+
+    swarm_coords[...] = coords
+    swarm_parent_cell_nums[...] = parent_cell_nums
+    field_parent_cell_nums[...] = parent_cell_nums
```

Comment:
Check these are actually the same in all/most cases

Reply:
OK so it turns out they aren't. We use a permuted plex. The permutation (of all points in the plex) is `parent_mesh._plex_renumbering`. It's a PETSc IS. If I generate a PETSc section with the global numbering for what is essentially DG[0] (i.e. one DoF per cell) then that section encodes the renumbering from the PETSc cell numbers to the Firedrake ones.

I guess what I need is an IS equivalent to the contents of that section. I could then invert it and apply it to the cell numbers we have in order to get the PETSc cell numbers. I am not sure how to do that.

Reply:
https://github.com/firedrakeproject/firedrake/blob/bae7e705755ae95cf4fc4efec488b1b6ec1b7955/firedrake/cython/mgimpl.pyx#L19

Reply:
Can we add a test that uses both `swarm.setPointCoordinates` and our new drop in replacement to make sure we get the same numbering?

Reply:
OK this is done now thanks to @ksagiyam who pointed out the easy way of getting the permutation I needed. I have also extended the swarm tests to check that we get the same that PETSc would.

Along the way I worked out that we can also remove points in the cell nodes for free so we no longer have to separately do that.